### ISSUE TITLE
Captured identifiers in SQL strings

### ISSUE THREAD

**Issue by gwenn** [rusqlite/rusqlite#1253]
I am not sure it is feasible but:
With the [Tcl interface to SQLite](https://sqlite.org/tclsqlite.html), you can do:
```tcl
db1 eval {INSERT INTO t1 VALUES(5,$bigstring)}
```
where `$bigstring` is a tcl variable.

And Rust [1.58.0](https://blog.rust-lang.org/2022/01/13/Rust-1.58.0.html#captured-identifiers-in-format-strings) supports this syntax:
```rust
println!("Hello, {person}!");
```
Is this feature only available to the standard library ?
If not, we should be able to improve `rusqlite` API !

**Comment by thomcc** [comment]
We can do this via a proc macro, but it requires parsing SQL.

**Comment by gwenn** [comment]
Ok, I think we can use [sqlite3-parser](https://crates.io/crates/sqlite3-parser) without modification.
Because, by default, the SQLite3 syntax accepts this kind of parameter:
https://sqlite.org/lang_expr.html#varparam
> A dollar-sign followed by an identifier name also holds a spot for a named parameter with the name $AAAA. The identifier name in this case can include one or more occurrences of "::" and a suffix enclosed in "(...)" containing any text at all. This syntax is the form of a variable name in the [Tcl programming language](http://www.tcl-lang.org/). The presence of this syntax results from the fact that SQLite is really a [Tcl extension](https://sqlite.org/tclsqlite.html) that has escaped into the wild.
And we can [already](https://github.com/gwenn/lemon-rs/blob/master/src/lexer/sql/test.rs#L37-L47) iterate on parameters.
But I am not an expert in proc macro...

**Comment by gwenn** [comment]
Is [fstrings-rs](https://github.com/danielhenrymantilla/fstrings-rs/blob/master/src/proc_macro/mod.rs) a good start ?

**Comment by thomcc** [comment]
The one I'd use as reference is https://github.com/matklad/xshell/blob/master/xshell-macros/src/lib.rs. It's simpler, closer to what we'd be doing, and is compatible with the standard captured identifiers stuff.

**Comment by gwenn** [comment]
SQLite parser rejects expression:
```sql
sqlite> .parameter init
sqlite> .parameter set $test 'hello'
sqlite> SELECT $test;
hello
sqlite> .parameter set $x.y 'abc'
sqlite> SELECT $x.y;
Parse error: near ".": syntax error
  SELECT $x.y;
           ^--- error here
```

**Comment by gwenn** [comment]
Initial draft [here](https://github.com/gwenn/rusqlite/tree/captured_identifiers).

**Comment by thomcc** [comment]
Doesn't look bad. There are holidays in the US this week so I won't really be able to look until next weekend though.

**Comment by gwenn** [comment]
I guess the following errors are related to macro hygiene but I am not sure:
```
rusqlite % cargo test
   Compiling rusqlite v0.29.0 (/Users/gwen/Rust/rusqlite)
error[E0425]: cannot find value `name` in this scope
    --> src/lib.rs:224:9
     |
224  |         $crate::__bind!(stmt $sql);
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope
...
2122 |         let mut stmt = prepare_and_bind!(db, "SELECT $name, $age;");
     |                        -------------------------------------------- in this macro invocation
     |
     = note: this error originates in the macro `$crate::__bind` which comes from the expansion of the macro `prepare_and_bind` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find value `age` in this scope
    --> src/lib.rs:224:9
     |
224  |         $crate::__bind!(stmt $sql);
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope
...
2122 |         let mut stmt = prepare_and_bind!(db, "SELECT $name, $age;");
     |                        -------------------------------------------- in this macro invocation
     |
     = note: this error originates in the macro `$crate::__bind` which comes from the expansion of the macro `prepare_and_bind` (in Nightly builds, run with -Z macro-backtrace for more info)
```
https://github.com/gwenn/rusqlite/blob/0e369ba878c4e5915049fdfed931c6ff56af876d/src/lib.rs#L2118-L2122
```rust
    fn prepare_and_bind() -> Result<()> {
        let db = Connection::open_in_memory()?;
        let name = "Lisa";
        let age = 8;
        let mut stmt = prepare_and_bind!(db, "SELECT $name, $age;");
```
`cargo expand --tests --lib`
```rust
    fn prepare_and_bind() -> Result<()> {
        let db = Connection::open_in_memory()?;
        let name = "Lisa";
        let age = 8;
        let mut stmt = {
            let mut stmt = db.prepare("SELECT $name, $age;")?;
            stmt.raw_bind_parameter(1, &name)?;
            stmt.raw_bind_parameter(2, &age)?;
            stmt
        };

```

**Comment by gwenn** [comment]
It seems to be related to `Span`:
https://doc.rust-lang.org/proc_macro/struct.Ident.html#method.new
> Span::call_site() vs Span::def_site()

And https://github.com/matklad/xshell/blob/af75dd49c3bb78887aa29626b1db0b9d0a202946/xshell-macros/src/lib.rs#L101

**Comment by gwenn** [comment]
Fixed with `rusqlite` version 0.30.0
